### PR TITLE
Feature: Add support for `either`

### LINKED ISSUE #455: Add `Either` support
[`either`](https://docs.rs/either/latest/either/) is a simple crate that exposes `Either` monad to Rust users. `Either` is an enum of `Left(T)` and `Right(U)`, allowing storing *either* of two types (hence the name). Other types widely used in Rust are variants of `Either`, including `Result<T, E> = Either<T, E>` and `Option<T> = Either<T, ()>`.

By providing blanket implementations of `TryFrom<Robj>` and `IntoRobj` for `Either<T, U>` where both `T` and `U` are `TryFrom<Robj>`/`IntoRobj`, we can add support for interesting scenarios in `extendr`.

One particular thing I have in mind is the problem of `1:3` vs `c(1, 2, 3)`. From R perspective, the first one is an `ALTINTEGER`, while the second one is `REAL`. There is also a variant of `c(1L, 2L, 3L)`, an `INTEGER`, but nobody writes like this. The issue here is if we want to allow passing any of these values to extendr, we should accept `Robj` and perform some pattern matching directly in the code. This reduces the safety guarantees of the type system (since we accept *anything*) and also adds extra boilerplate.

With `Either` support, we can safely write the following code (based on my PoC implementation):

``` r
rextendr::rust_source(
    code = "
    #[extendr(use_try_from = true)]
    fn get_sum(x : Either<Integers, Doubles>) -> Rfloat {
        match x {
            Left(ints) => ints.iter().sum::<Rint>().into(),
            Right(dbls) => dbls.iter().sum()
        }
    }
    ",
    extendr_deps = list(
        `extendr-api` = list(
            path = "...\\extendr\\extendr-api"
        )
    )
)
#> ℹ build directory: 'C:/Users/.../AppData/Local/Temp/Rtmpu6Swt3/file31b061521279'
#> ✔ Writing 'C:/Users/.../AppData/Local/Temp/Rtmpu6Swt3/file31b061521279/target/extendr_wrappers.R'

1:3 |> get_sum()
#> [1] 6
c(1, 2, 3) |> get_sum()
#> [1] 6
c(1L, 2L, 3L) |> get_sum()
#> [1] 6
```

<sup>Created on 2023-01-22 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by Ilia-Kosenkov**
- Provide an implementation of `TryFrom<&Robj>` for types `Either<T, U>` where `T` and `U` are `TryFrom<&Robj>`
  - When matching, first check `T` and if `try_from` fails
  - Check `U` 
- Provide an implementation of `IntoRobj` for types `Either<T, U>` where `T` and `U` are `IntoRobj`

Closes #455

### CODE REVIEW

File: extendr-api/tests/optional/either/either_try_from_tests.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+use either::Either::{self, Left, Right};
```
Comment (multimeric) [inline extendr-api/tests/optional/either/either_try_from_tests.rs:1]: `either_try_from_robj` as the filename? We can drop the `_test` suffix I guess.

File: extendr-api/tests/optional/either/either_try_from_tests.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+{
+    test! {
+        let val = eval_string(src)?;
+        let val = Either::<TLeft, TRight>::try_from(&val)?;
+        assert_eq!(val, Left(left));
+    }
+}
+
+#[rstest]
+#[case("1L", Rint::from(1i32), "")]
```
Comment (multimeric) [inline extendr-api/tests/optional/either/either_try_from_tests.rs:27]: Repeated, as above.

File: tests/extendrtests/tests/testthat/test-optional-either.R [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    },
+    patrick::cases(
+        LOGICAL = list(input = c(TRUE, FALSE, NA)),
+        CHARACTER = list(input = letters),
+        COMPLEX = list(input = 1:10 + 1i * 11:20),
+        `NULL` = list(input = NULL),
+        `DATA.FRAME` = list(input = mtcars),
+        LIST = list(input = lapply(1:10, identity))
+    )
+)
```
Comment (multimeric) [inline tests/extendrtests/tests/testthat/test-optional-either.R:31]: Linebreak

**Review by multimeric — CHANGES_REQUESTED** [review]
* [ ] Update changelog

**Comment by multimeric** [comment]
I still don't really get the point of this feature, but since it's feature locked it won't hurt anyone who doesn't want it, so I will be happy to merge once everything is resolved.

File: extendr-api/src/optional/either.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+use crate::{Error, Robj};
+use either::Either::{self, Left, Right};
+
+impl<'a, TLeft, TRight> TryFrom<&'a Robj> for Either<TLeft, TRight>
```
Comment (CGMossa) [inline extendr-api/src/optional/either.rs:4]: Maybe this call them `Left` and `Right`?

**Comment by CGMossa** [comment]
Suggested edit:
```diff
diff --git a/extendr-api/Cargo.toml b/extendr-api/Cargo.toml
index 028f39e5..11e0dcb4 100644
--- a/extendr-api/Cargo.toml
+++ b/extendr-api/Cargo.toml
@@ -18,10 +18,10 @@ repository = "https://github.com/extendr/extendr"
 libR-sys = "0.3.0"
 extendr-macros = { path = "../extendr-macros", version = "0.3.1" }
 extendr-engine = { path = "../extendr-engine", version = "0.3.1" }
-either = { version = "1.8.1", optional = true }
-ndarray = { version = "0.15.3", optional = true }
 lazy_static = "1.4"
 paste = "1.0.5"
+either = { version = "1.8.1", optional = true }
+ndarray = { version = "0.15.3", optional = true }
 serde = { version = "1.0", features = ["derive"], optional = true }
 num-complex = { version = "0.4", optional = true }
 libc = { version = "0.2", optional = true }

```

**Comment by CGMossa** [comment]
Suggested edit:
```diff
diff --git a/extendr-api/Cargo.toml b/extendr-api/Cargo.toml
index 028f39e5..6d8f2cdf 100644
--- a/extendr-api/Cargo.toml
+++ b/extendr-api/Cargo.toml
@@ -18,10 +18,10 @@ repository = "https://github.com/extendr/extendr"
 libR-sys = "0.3.0"
 extendr-macros = { path = "../extendr-macros", version = "0.3.1" }
 extendr-engine = { path = "../extendr-engine", version = "0.3.1" }
-either = { version = "1.8.1", optional = true }
-ndarray = { version = "0.15.3", optional = true }
 lazy_static = "1.4"
 paste = "1.0.5"
+either = { version = "1.8.1", optional = true }
+ndarray = { version = "0.15.3", optional = true }
 serde = { version = "1.0", features = ["derive"], optional = true }
 num-complex = { version = "0.4", optional = true }
 libc = { version = "0.2", optional = true }
@@ -30,7 +30,7 @@ libc = { version = "0.2", optional = true }
 rstest = "0.16.0"
 
 [features]
-default = []
+default = ["either"]
 
 # libc is needed to allocate a DevDesc (c.f., https://bugs.r-project.org/show_bug.cgi?id=18292)
 graphics = ["libc"]

```

File: extendr-api/tests/optional/either/either_robj_from_tests.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
+use either::Either::{self, Left, Right};
+use extendr_api::prelude::*;
+use rstest::rstest;
+
+#[rstest]
+#[case("1L", Rint::from(1i32), "")]
+#[case("1.5", Rfloat::from(1.5), Rint::default())]
+#[case("\"string\"", "string", false)]
+#[case("TRUE", true, 0)]
+#[case("1 + 2i", Rcplx::new(1.0, 2.0), 0)]
```
Comment (CGMossa) [inline extendr-api/tests/optional/either/either_robj_from_tests.rs:10]: What happens if you use the true "omit me" type in Rust i.e. `()` for `_TRight` here?

**Comment by CGMossa** [comment]
Suggested edit:
```diff
diff --git a/tests/extendrtests/src/rust/Cargo.toml b/tests/extendrtests/src/rust/Cargo.toml
index 6bf147a1..a1344bd7 100644
--- a/tests/extendrtests/src/rust/Cargo.toml
+++ b/tests/extendrtests/src/rust/Cargo.toml
@@ -15,7 +15,7 @@ crate-type = ["staticlib"]
 
 [dependencies]
 extendr-api = { version = "*", features = ["graphics", "ndarray", "either"] }
-either = '*'
+either = '1'
 
 [patch.crates-io]
 ## This is configured to work with RStudio features.

```

**Comment by CGMossa** [comment]
Suggested edit:
```diff
diff --git a/extendr-api/src/optional/either.rs b/extendr-api/src/optional/either.rs
index 931b390c..28ac0ef8 100644
--- a/extendr-api/src/optional/either.rs
+++ b/extendr-api/src/optional/either.rs
@@ -8,6 +8,8 @@ where
 {
     type Error = Error;
 
+    /// Returns the first type that matches the provided `Robj`, starting from
+    /// `Left`-type, and if that fails, then the `Right`-type is converted.
     fn try_from(value: &'a Robj) -> Result<Self, Self::Error> {
         match TLeft::try_from(value) {
             Ok(left) => Ok(Left(left)),

```

**Comment by multimeric** [comment]
@CGMossa it would be ideal if you used the GitHub suggestion feature where possible, rather than posting diffs: 
image

**Comment by CGMossa** [comment]
So the positing diffs comes from using githubs pull request extension for
vscode. I had hoped it did more than just post diffs.
Sorry about that I'll try something elske next time.

**Comment by Ilia-Kosenkov** [comment]
No worries, I can decipher your suggestions anyway :)

File: extendr-api/tests/optional/either/either_robj_from_tests.rs (hunk 5 above)
Comment (Ilia-Kosenkov) [reply extendr-api/tests/optional/either/either_robj_from_tests.rs]: No longer applicable

**Review by CGMossa — APPROVED** [review]

File: extendr-api/src/optional/either.rs (hunk 4 above)
Comment (Ilia-Kosenkov) [reply extendr-api/src/optional/either.rs:4]: `Either` uses `T` and `R` for their names, so I decided to go the same way.

Comment (CGMossa) [reply extendr-api/src/optional/either.rs:4]: Either use left and right, or R and L, but this current one is a little ugly. Even T and R is better.

**Comment by CGMossa** [comment]
Can we consider making either feature a default one please?

**Comment by Ilia-Kosenkov** [comment]
How about we first leave it behind a feature and maybe revisit it later? Some of us see a lot of value in it, others are more hesitant.

**Comment by multimeric** [comment]
We can merge this, I really don't mind! If it's passing all tests and it's feature locked I see no downside in it being in the release.

**Comment by CGMossa** [comment]
Me too.. I vote merge.

**Comment by multimeric** [comment]
If Ilia agrees, I'll include this in 0.4.0.

By the way, one of the reasons I wanted to keep this feature locked was because this pull in another crate and I want users to be able to avoid this if they want.

**Comment by CGMossa** [comment]
Yes but it is a fairly small crate, that honestly will be supported in rust
natively, if it wasn't for the fact that support for.. I cannot remember it
now, but it is needed before standardization.

**Comment by CGMossa** [comment]
Please add `"either"` to this line
```toml
[package.metadata.docs.rs]
features = [
```
in `extendr-api/Cargo.toml`.

This is due to #494 (will conflict with #495, sorry :( )

**Comment by Ilia-Kosenkov** [comment]
I am OK with merging it whenever you feel appropriate. @multimeric , I think you need to hit something here to accept the changes that you've requested.

**Comment by multimeric** [comment]
As @CGMossa said, you need to fix the docs rs features.

**Comment by Ilia-Kosenkov** [comment]
Done

**Review by multimeric — APPROVED** [review]
Looks good! You've managed to distil this PR into what seems like a much shorter one which is quite easy to follow.

**Pull request merged**